Add support for custom preprocessor definitions

stijnh opened the pull request:
By default, kernel tuner inserts the tunable parameters as preprocessor parameters into the source code. However thanks to the cupy backed, these parameters can now also be inserted using template arguments, meaning the preprocessor definitions are no longer necessary. It would nice if there was a way to disable them since they can cause conflicts with external libraries (for instance, CUB uses the variable "block_size" in a few places).

This PR adds a new keyword argument `defines` to `tune_kernel` (and `run_kernel`) that allows customizing the behavior of kernel tuner around these preprocessor definitions. 

Examples:
* if `defines=None`, the behavior is identical to how it is today
* if `defines=dict()`, no preprocessor definitions are inserted, not even the default ones
* if `defines=dict(floating_type="float")`, insert `#define floating_type float`
* if `defines=dict(block_size=lambda p: p["block_size"])`, define only `block_size` 
* if `defines=dict(twice_block_size=lambda p: 2 * p["block_size"])`, insert `twice_block_size` with the value of `2 * block_size`.

Unresolved issues:
- [x] Should this keyword be called `defines` or `definitions`?
- [x] Do we need more unit-tests?
- [x] Is this behavior intuitive? An alternative would be to have a flag `disable_default_defines=True` to disable the default definitions, but this would add yet another keyword argument :-(

benvanwerkhoven commented:
Hi Stijn!

Thanks for this! I can definitely see this being useful in certain cases. ``defines`` seems like a fine name for this. Also passing an empty dict to disable default behavior is similar to how we disable other defaults in Kernel Tuner, for example with grid divisors. 

I don't fully understand yet how the parameters are inserted through template arguments in the absence of defines. If you look at this example:
https://github.com/benvanwerkhoven/kernel_tuner/blob/master/examples/cuda-c%2B%2B/vector_add_blocksize.py
We insert the parameters as template arguments but these currently only have meaningful values because of the preprocessor defines.

I've added an example to this branch just now, which currently isn't working:
https://github.com/stijnh/kernel_tuner/blob/custom-defines/examples/cuda-c%2B%2B/vector_add_defines.py

I believe the question now is how to enable the insertion of tunable parameter values into template arguments. I don't know if you already had something in mind for this, but my first guess would be to just allow someone to pass
``kernel_name = "vector_add<block_size_x>"`` and then use a util function like [replace_param_occurances](https://github.com/benvanwerkhoven/kernel_tuner/blob/c706922288e3234b3b5fe88c6ad37e1719b6c717/kernel_tuner/util.py#L583) to replace the tunable parameters with their value at this configuration. This would be a very easy solution in my opinion. Do you agree or did you have something else in mind?

stijnh commented:
Indeed, I had not considered that disabling the preprocessor defines would make it impossible to pass the arguments as template parameter. I have fixed this in `prepare_kernel_string` so that it now also replace the parameters in the kernel name.

I have also modified `replace_param_occurrences` such that it first tokenizes the input instead of simply performing a string replace.

I feel this branch is ready for merging!

benvanwerkhoven commented:
Thanks a lot Stijn!
